File: packages/java-sdk/src/main/java/com/turbodocx/HttpClient.java
package com.turbodocx;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.TypeAdapter;
import com.google.gson.stream.JsonReader;
import com.google.gson.stream.JsonToken;
import com.google.gson.stream.JsonWriter;
import okhttp3.*;

import java.io.IOException;
import java.lang.reflect.Type;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.concurrent.TimeUnit;

/**
 * File type detection result
 */
class FileTypeInfo {
    public final String mimeType;
    public final String extension;

    public FileTypeInfo(String mimeType, String extension) {
        this.mimeType = mimeType;
        this.extension = extension;
    }
}

/**
 * HTTP client wrapper for TurboDocx API
 */
public class HttpClient {
    private static final String DEFAULT_BASE_URL = "https://api.turbodocx.com";
    private static final MediaType JSON = MediaType.parse("application/json; charset=utf-8");

    /**
     * Detect file type from magic bytes
     */
    public static FileTypeInfo detectFileType(byte[] fileBytes) {
        if (fileBytes == null || fileBytes.length < 4) {
            return new FileTypeInfo("application/octet-stream", "bin");
        }

        // PDF: %PDF (0x25 0x50 0x44 0x46)
        if (fileBytes[0] == 0x25 && fileBytes[1] == 0x50 && fileBytes[2] == 0x44 && fileBytes[3] == 0x46) {
            return new FileTypeInfo("application/pdf", "pdf");
        }

        // ZIP-based formats (DOCX, PPTX): starts with PK (0x50 0x4B)
        if (fileBytes[0] == 0x50 && fileBytes[1] == 0x4B) {
            int headerLen = Math.min(fileBytes.length, 2000);
            String header = new String(Arrays.copyOf(fileBytes, headerLen), StandardCharsets.UTF_8);

            // PPTX contains 'ppt/' in the ZIP structure
            if (header.contains("ppt/")) {
                return new FileTypeInfo(
                    "application/vnd.openxmlformats-officedocument.presentationml.presentation",
                    "pptx"
                );
            }

            // DOCX contains 'word/' in the ZIP structure
            if (header.contains("word/")) {
                return new FileTypeInfo(
                    "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
                    "docx"
                );
            }

            // Default to DOCX for unknown ZIP
            return new FileTypeInfo(
                "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
                "docx"
            );
        }

        // Unknown file type
        return new FileTypeInfo("application/octet-stream", "bin");
    }

    private final OkHttpClient client;
    private final String baseUrl;
    private final String apiKey;
    private final String accessToken;
    private final String orgId;
    private final String senderEmail;
    private final String senderName;
    private final Gson gson;

    public HttpClient(String baseUrl, String apiKey, String accessToken, String orgId, String senderEmail, String senderName) {
        this(baseUrl, apiKey, accessToken, orgId, senderEmail, senderName, 60, 120, 60);
    }

    public HttpClient(String baseUrl, String apiKey, String accessToken, String orgId, String senderEmail, String senderName,
                       int connectTimeoutSeconds, int readTimeoutSeconds, int writeTimeoutSeconds) {
        this.client = new OkHttpClient.Builder()
                .connectTimeout(connectTimeoutSeconds, TimeUnit.SECONDS)
                .readTimeout(readTimeoutSeconds, TimeUnit.SECONDS)
                .writeTimeout(writeTimeoutSeconds, TimeUnit.SECONDS)
                .build();
        this.baseUrl = baseUrl != null ? baseUrl.replaceAll("/$", "") : DEFAULT_BASE_URL;
        this.apiKey = apiKey;
        this.accessToken = accessToken;
        this.orgId = orgId;
        this.senderEmail = senderEmail;
        this.senderName = senderName;
        this.gson = new GsonBuilder()
                .registerTypeAdapter(int.class, new FlexIntAdapter())
                .registerTypeAdapter(Integer.class, new FlexIntAdapter())
                .create();
    }

    /**
     * Get the sender email configuration
     */
    public String getSenderEmail() {
        return senderEmail;
    }

    /**
     * Get the sender name configuration
     */
    public String getSenderName() {
        return senderName;
    }

    /**
     * Get the underlying OkHttpClient for inspection (package-private).
     */
    OkHttpClient getOkHttpClient() {
        return client;
    }

    /**
     * Shut down the underlying OkHttpClient's connection pool and dispatcher.
     */
    public void close() {
        client.dispatcher().executorService().shutdown();
        client.connectionPool().evictAll();
        okhttp3.Cache cache = client.cache();
        if (cache != null) {
            try { cache.close(); } catch (IOException ignored) { }
        }
    }

    public <T> T get(String path, Class<T> responseClass) throws IOException {
        Request request = new Request.Builder()
                .url(baseUrl + path)
                .headers(buildHeaders())
                .get()
                .build();

        return execute(request, responseClass);
    }

    public byte[] getRaw(String path) throws IOException {
        Request request = new Request.Builder()
                .url(baseUrl + path)
                .headers(buildHeaders())
                .get()
                .build();

        try (Response response = client.newCall(request).execute()) {
            if (!response.isSuccessful()) {
                handleError(response);
            }
            return response.body() != null ? response.body().bytes() : new byte[0];
        }
    }

    public <T> T get(String path, Map<String, Object> queryParams, Class<T> responseClass) throws IOException {
        String url = buildUrl(path, queryParams);
        Request request = new Request.Builder()
                .url(url)
                .headers(buildHeaders())
                .get()
                .build();

        return execute(request, responseClass);
    }

    public <T> T get(String path, Map<String, Object> queryParams, Type responseType) throws IOException {
        String url = buildUrl(path, queryParams);
        Request request = new Request.Builder()
                .url(url)
                .headers(buildHeaders())
                .get()
                .build();

        return execute(request, responseType);
    }

    public <T> T get(String path, Type responseType) throws IOException {
        Request request = new Request.Builder()
                .url(baseUrl + path)
                .headers(buildHeaders())
                .get()
                .build();

        return execute(request, responseType);
    }

    public <T> T post(String path, Object body, Class<T> responseClass) throws IOException {
        RequestBody requestBody = body != null
                ? RequestBody.create(gson.toJson(body), JSON)
                : RequestBody.create("{}", JSON);

        Request request = new Request.Builder()
                .url(baseUrl + path)
                .headers(buildHeaders())
                .post(requestBody)
                .build();

        return execute(request, responseClass);
    }

    public <T> T post(String path, Object body, Type responseType) throws IOException {
        RequestBody requestBody = body != null
                ? RequestBody.create(gson.toJson(body), JSON)
                : RequestBody.create("{}", JSON);

        Request request = new Request.Builder()
                .url(baseUrl + path)
                .headers(buildHeaders())
                .post(requestBody)
                .build();

        return execute(request, responseType);
    }

    public <T> T patch(String path, Object body, Class<T> responseClass) throws IOException {
        RequestBody requestBody = body != null
                ? RequestBody.create(gson.toJson(body), JSON)
                : RequestBody.create("{}", JSON);

        Request request = new Request.Builder()
                .url(baseUrl + path)
                .headers(buildHeaders())
                .patch(requestBody)
                .build();

        return execute(request, responseClass);
    }

    public <T> T patch(String path, Object body, Type responseType) throws IOException {
        RequestBody requestBody = body != null
                ? RequestBody.create(gson.toJson(body), JSON)
                : RequestBody.create("{}", JSON);

        Request request = new Request.Builder()
                .url(baseUrl + path)
                .headers(buildHeaders())
                .patch(requestBody)
                .build();

        return execute(request, responseType);
    }

    /**
     * PATCH with a pre-serialized JSON string body.
     * Use this when the body has already been serialized (e.g., to preserve explicit nulls
     * that default Gson would drop).
     */
    public <T> T patchRawJson(String path, String jsonBody, Type responseType) throws IOException {
        RequestBody requestBody = jsonBody != null
                ? RequestBody.create(jsonBody, JSON)
                : RequestBody.create("{}", JSON);

        Request request = new Request.Builder()
                .url(baseUrl + path)
                .headers(buildHeaders())
                .patch(requestBody)
                .build();

        return execute(request, responseType);
    }

    /**
     * POST with a pre-serialized JSON string body.
     * Use this when the body has already been serialized (e.g., to preserve an explicit null
     * that default Gson would drop — see {@code TurboQuote.addLineItems} custom line items).
     */
    public <T> T postRawJson(String path, String jsonBody, Type responseType) throws IOException {
        RequestBody requestBody = jsonBody != null
                ? RequestBody.create(jsonBody, JSON)
                : RequestBody.create("{}", JSON);

        Request request = new Request.Builder()
                .url(baseUrl + path)
                .headers(buildHeaders())
                .post(requestBody)
                .build();

        return execute(request, responseType);
    }

    public <T> T delete(String path, Class<T> responseClass) throws IOException {
        Request request = new Request.Builder()
                .url(baseUrl + path)
                .headers(buildHeaders())
                .delete()
                .build();

        return execute(request, responseClass);
    }

    public <T> T postFormData(String path, MultipartBody body, Class<T> responseClass) throws IOException {
        Request request = new Request.Builder()
                .url(baseUrl + path)
                .headers(buildHeaders())
                .post(body)
                .build();

        return execute(request, responseClass);
    }

    public <T> T postFormData(String path, MultipartBody body, Type responseType) throws IOException {
        Request request = new Request.Builder()
                .url(baseUrl + path)
                .headers(buildHeaders())
                .post(body)
                .build();

        return execute(request, responseType);
    }

    public <T> T patchFormData(String path, MultipartBody body, Class<T> responseClass) throws IOException {
        Request request = new Request.Builder()
                .url(baseUrl + path)
                .headers(buildHeaders())
                .patch(body)
                .build();

        return execute(request, responseClass);
    }

    public <T> T patchFormData(String path, MultipartBody body, Type responseType) throws IOException {
        Request request = new Request.Builder()
                .url(baseUrl + path)
                .headers(buildHeaders())
                .patch(body)
                .build();

        return execute(request, responseType);
    }

    /**
     * Upload file from bytes
     */
    public <T> T uploadFile(String path, byte[] file, String fileName, Map<String, String> formData, Class<T> responseClass) throws IOException {
        // Auto-detect filename from content if not provided
        if (fileName == null || fileName.isEmpty()) {
            FileTypeInfo detected = detectFileType(file);
            fileName = "document." + detected.extension;
        }

        // Detect MIME type from content
        FileTypeInfo detected = detectFileType(file);

        MultipartBody.Builder builder = new MultipartBody.Builder()
                .setType(MultipartBody.FORM)
                .addFormDataPart("file", fileName,
                        RequestBody.create(file, MediaType.parse(detected.mimeType)));

        for (Map.Entry<String, String> entry : formData.entrySet()) {
            builder.addFormDataPart(entry.getKey(), entry.getValue());
        }

        Request request = new Request.Builder()
                .url(baseUrl + path)
                .headers(buildHeaders())
                .post(builder.build())
                .build();

        return execute(request, responseClass);
    }

    /**
     * Upload file from file path (using java.nio.file.Path)
     */
    public <T> T uploadFile(String path, Path filePath, Map<String, String> formData, Class<T> responseClass) throws IOException {
        byte[] fileBytes = Files.readAllBytes(filePath);
        String fileName = filePath.getFileName().toString();
        return uploadFile(path, fileBytes, fileName, formData, responseClass);
    }

    /**
     * Upload file from file path (using String path)
     */
    public <T> T uploadFilePath(String path, String filePath, Map<String, String> formData, Class<T> responseClass) throws IOException {
        return uploadFile(path, Paths.get(filePath), formData, responseClass);
    }

    private <T> T execute(Request request, Class<T> responseClass) throws IOException {
        return execute(request, (Type) responseClass);
    }

    @SuppressWarnings("unchecked")
    private <T> T execute(Request request, Type responseType) throws IOException {
        try (Response response = client.newCall(request).execute()) {
            if (!response.isSuccessful()) {
                handleError(response);
            }

            String responseBody = response.body() != null ? response.body().string() : "";

            // Parse to JsonElement for smart unwrap + normalization
            JsonElement parsed = gson.fromJson(responseBody, JsonElement.class);

            // Smart unwrapping: if response has ONLY "data" key, extract it
            // This handles backend responses that wrap data in { "data": { ... } }
            if (parsed != null && parsed.isJsonObject()) {
                JsonObject jsonObj = parsed.getAsJsonObject();
                if (jsonObj.has("data") && jsonObj.size() == 1) {
                    parsed = jsonObj.get("data");
                }
            }

            // Normalize MySQL type coercion (tinyint booleans, decimal strings)
            parsed = ResponseNormalizer.normalize(parsed);

            // Deserialize to target type
            return gson.fromJson(parsed, responseType);
        }
    }

    private void handleError(Response response) throws IOException {
        String body = response.body() != null ? response.body().string() : "";
        String message = "API Error";
        String code = null;

        try {
            JsonObject json = gson.fromJson(body, JsonObject.class);
            if (json != null) {
                // Check both "message" and "error" fields (backend uses both)
                if (json.has("message")) {
                    message = json.get("message").getAsString();
                } else if (json.has("error")) {
                    message = json.get("error").getAsString();
                }
                if (json.has("code")) {
                    code = json.get("code").getAsString();
                }
            }
        } catch (Exception e) {
            // Use default message
        }

        // Throw specific exception based on status code
        switch (response.code()) {
            case 400:
                throw new TurboDocxException.ValidationException(message, code);
            case 401:
                throw new TurboDocxException.AuthenticationException(message, code);
            case 403:
                throw new TurboDocxException.AuthorizationException(message, code);
            case 404:
                throw new TurboDocxException.NotFoundException(message, code);
            case 409:
                throw new TurboDocxException.ConflictException(message, code);
            case 429:
                throw new TurboDocxException.RateLimitException(message, code);
            default:
                throw new TurboDocxException(message, response.code(), code);
        }
    }

    /**
     * Gson TypeAdapter that handles both numeric (0/1) and boolean (true/false)
     * values for int fields. MySQL returns tinyint(1) as 0/1 but some contexts
     * may send true/false booleans.
     */
    private static class FlexIntAdapter extends TypeAdapter<Integer> {
        @Override
        public void write(JsonWriter out, Integer value) throws IOException {
            if (value == null) {
                out.nullValue();
            } else {
                out.value(value);
            }
        }

        @Override
        public Integer read(JsonReader in) throws IOException {
            JsonToken token = in.peek();
            switch (token) {
                case NUMBER:
                    return in.nextInt();
                case BOOLEAN:
                    return in.nextBoolean() ? 1 : 0;
                case NULL:
                    in.nextNull();
                    return 0;
                default:
                    throw new IOException("Expected NUMBER or BOOLEAN but was " + token);
            }
        }
    }

    /**
     * Build a URL with query parameters.
     * Supports String values, List&lt;String&gt; values (repeated keys), and null values (skipped).
     */
    private String buildUrl(String path, Map<String, Object> queryParams) {
        if (queryParams == null || queryParams.isEmpty()) {
            return baseUrl + path;
        }

        HttpUrl.Builder urlBuilder = HttpUrl.parse(baseUrl + path).newBuilder();
        for (Map.Entry<String, Object> entry : queryParams.entrySet()) {
            String key = entry.getKey();
            Object value = entry.getValue();
            if (value == null) continue;
            if (value instanceof List) {
                @SuppressWarnings("unchecked")
                List<String> list = (List<String>) value;
                for (String item : list) {
                    urlBuilder.addQueryParameter(key, item);
                }
            } else {
                urlBuilder.addQueryParameter(key, value.toString());
            }
        }
        return urlBuilder.build().toString();
    }

    private Headers buildHeaders() {
        Headers.Builder builder = new Headers.Builder();

        // API key is sent as Bearer token (backend expects Authorization header)
        if (accessToken != null && !accessToken.isEmpty()) {
            builder.add("Authorization", "Bearer " + accessToken);
        } else if (apiKey != null && !apiKey.isEmpty()) {
            builder.add("Authorization", "Bearer " + apiKey);
        }

        // Organization ID header (required by backend)
        if (orgId != null && !orgId.isEmpty()) {
            builder.add("x-rapiddocx-org-id", orgId);
        }

        return builder.build();
    }
}

File: packages/ruby-sdk/lib/turbodocx_sdk/http_client.rb
# frozen_string_literal: true

require "net/http"
require "uri"
require "json"
require "securerandom"
require "stringio"
require_relative "errors"
require_relative "response_normalizer"

module TurboDocxSdk
  # Low-level HTTP client for the TurboDocx API.
  #
  # Handles authentication, error mapping, response unwrapping, and
  # the response normalizer pipeline.
  class HttpClient
    DEFAULT_BASE_URL = "https://api.turbodocx.com"

    # @param api_key [String, nil]
    # @param access_token [String, nil]
    # @param base_url [String, nil]
    # @param org_id [String, nil]
    # @param sender_email [String, nil]
    # @param sender_name [String, nil]
    # @param skip_sender_validation [Boolean]
    # @raise [AuthenticationError] if no API key or access token is provided
    # @raise [ValidationError] if senderEmail is missing and validation is not skipped
    def initialize(
      api_key: nil,
      access_token: nil,
      base_url: nil,
      org_id: nil,
      sender_email: nil,
      sender_name: nil,
      skip_sender_validation: false
    )
      @api_key = api_key || ENV["TURBODOCX_API_KEY"]
      @access_token = access_token
      @base_url = base_url || ENV["TURBODOCX_BASE_URL"] || DEFAULT_BASE_URL
      @org_id = org_id || ENV["TURBODOCX_ORG_ID"]
      @sender_email = sender_email || ENV["TURBODOCX_SENDER_EMAIL"]
      @sender_name = sender_name || ENV["TURBODOCX_SENDER_NAME"]

      unless @api_key || @access_token
        raise AuthenticationError, "API key or access token is required"
      end

      unless @sender_email || skip_sender_validation
        raise ValidationError,
              'senderEmail is required. This email will be used as the reply-to address for signature requests. ' \
              'Without it, emails will default to "API Service User via TurboSign".'
      end
    end

    # @return [Hash] sender email and name configuration
    def sender_config
      { "senderEmail" => @sender_email, "senderName" => @sender_name }
    end

    # GET request.
    # @param path [String]
    # @param params [Hash, nil] query params (values may be String or Array<String>)
    # @return [Object] parsed, unwrapped, normalized response
    # @raise [AuthenticationError] on 401 response
    # @raise [NotFoundError] on 404 response
    # @raise [RateLimitError] on 429 response
    # @raise [ValidationError] on 400 response
    # @raise [NetworkError] on connection failure
    def get(path, params = nil)
      url = build_url(path, params)
      request = Net::HTTP::Get.new(url)
      apply_headers(request)
      execute(request, url)
    end

    # POST request with JSON body.
    # @param path [String]
    # @param data [Hash, nil] request body
    # @return [Object] parsed, unwrapped, normalized response
    # @raise [ValidationError] on 400 response
    # @raise [AuthenticationError] on 401 response
    # @raise [NotFoundError] on 404 response
    # @raise [RateLimitError] on 429 response
    # @raise [NetworkError] on connection failure
    def post(path, data = nil)
      url = build_url(path)
      request = Net::HTTP::Post.new(url)
      apply_headers(request)
      if data
        request.body = JSON.generate(data)
      end
      execute(request, url)
    end

    # PATCH request with JSON body.
    # @param path [String]
    # @param data [Hash, nil] request body
    # @return [Object] parsed, unwrapped, normalized response
    # @raise [ValidationError] on 400 response
    # @raise [AuthenticationError] on 401 response
    # @raise [NotFoundError] on 404 response
    # @raise [RateLimitError] on 429 response
    # @raise [NetworkError] on connection failure
    def patch(path, data = nil)
      url = build_url(path)
      request = Net::HTTP::Patch.new(url)
      apply_headers(request)
      if data
        request.body = JSON.generate(data)
      end
      execute(request, url)
    end

    # DELETE request.
    # @param path [String]
    # @return [Object] parsed, unwrapped, normalized response
    # @raise [AuthenticationError] on 401 response
    # @raise [NotFoundError] on 404 response
    # @raise [RateLimitError] on 429 response
    # @raise [NetworkError] on connection failure
    def delete(path)
      url = build_url(path)
      request = Net::HTTP::Delete.new(url)
      apply_headers(request)
      execute(request, url)
    end

    # GET returning raw binary (for PDF downloads etc.).
    # @param path [String]
    # @param params [Hash, nil] query params
    # @return [String] raw response body bytes
    # @raise [AuthenticationError] on 401 response
    # @raise [NotFoundError] on 404 response
    # @raise [NetworkError] on connection failure
    def get_raw(path, params = nil)
      url = build_url(path, params)
      request = Net::HTTP::Get.new(url)
      apply_headers(request, content_type: false)
      execute_raw(request, url)
    end

    # POST with multipart/form-data.
    # @param path [String]
    # @param form_data [Hash] keys are field names, values are strings or
    #   Hashes with :io, :filename, :content_type for file parts.
    # @return [Object] parsed, unwrapped, normalized response
    # @raise [ValidationError] on 400 response
    # @raise [AuthenticationError] on 401 response
    # @raise [NetworkError] on connection failure
    def post_form_data(path, form_data)
      request_form_data("POST", path, form_data)
    end

    # PATCH with multipart/form-data.
    # @param path [String]
    # @param form_data [Hash] keys are field names, values are strings or
    #   Hashes with :io, :filename, :content_type for file parts.
    # @return [Object] parsed, unwrapped, normalized response
    # @raise [ValidationError] on 400 response
    # @raise [AuthenticationError] on 401 response
    # @raise [NetworkError] on connection failure
    def patch_form_data(path, form_data)
      request_form_data("PATCH", path, form_data)
    end

    # Upload a file (used by TurboSign).
    # @param api_path [String]
    # @param file [String, IO] file path or IO-like object
    # @param field_name [String] form field name for the file
    # @param additional_data [Hash, nil] extra form fields
    # @return [Object] parsed, unwrapped, normalized response
    # @raise [ValidationError] if the file input type is unsupported or on 400 response
    # @raise [AuthenticationError] on 401 response
    # @raise [NetworkError] on connection failure
    def upload_file(api_path, file, field_name: "file", additional_data: nil)
      parts = {}

      if file.is_a?(String)
        # file path
        content = File.binread(file)
        filename = File.basename(file)
        mime = detect_file_type(content)
        parts[field_name] = { io: StringIO.new(content), filename: filename, content_type: mime }
      elsif file.respond_to?(:read)
        content = file.read
        mime = detect_file_type(content)
        filename = additional_data&.dig("fileName") || "document.bin"
        parts[field_name] = { io: StringIO.new(content), filename: filename, content_type: mime }
      else
        raise ValidationError, "Unsupported file input type: #{file.class}"
      end

      if additional_data
        additional_data.each do |key, value|
          next if key == "fileName"
          parts[key.to_s] = value.is_a?(Hash) || value.is_a?(Array) ? JSON.generate(value) : value.to_s
        end
      end

      post_form_data(api_path, parts)
    end

    private

    def build_url(path, params = nil)
      url_str = "#{@base_url}#{path}"
      if params && !params.empty?
        query_parts = []
        params.each do |key, value|
          if value.is_a?(Array)
            value.each { |v| query_parts << "#{URI.encode_www_form_component(key)}=#{URI.encode_www_form_component(v)}" }
          else
            query_parts << "#{URI.encode_www_form_component(key)}=#{URI.encode_www_form_component(value)}"
          end
        end
        url_str += "?#{query_parts.join("&")}"
      end
      URI.parse(url_str)
    end

    def apply_headers(request, content_type: true)
      request["Content-Type"] = "application/json" if content_type
      if @access_token
        request["Authorization"] = "Bearer #{@access_token}"
      elsif @api_key
        request["Authorization"] = "Bearer #{@api_key}"
      end
      request["x-rapiddocx-org-id"] = @org_id if @org_id
    end

    def execute(request, url)
      response = http_request(request, url)
      handle_error_response(response) unless response.is_a?(Net::HTTPSuccess)

      content_type = response["content-type"]
      if content_type && content_type.include?("application/json")
        json_data = JSON.parse(response.body)
        ResponseNormalizer.normalize(smart_unwrap(json_data))
      else
        response.body
      end
    rescue TurboDocxError
      raise
    rescue StandardError => e
      raise NetworkError, "Network request failed: #{e.message}"
    end

    def execute_raw(request, url)
      response = http_request(request, url)
      handle_error_response(response) unless response.is_a?(Net::HTTPSuccess)
      response.body
    rescue TurboDocxError
      raise
    rescue StandardError => e
      raise NetworkError, "Network request failed: #{e.message}"
    end

    def http_request(request, url)
      http = Net::HTTP.new(url.host, url.port)
      http.use_ssl = url.scheme == "https"
      http.request(request)
    end

    def request_form_data(method, path, form_data)
      url = build_url(path)
      boundary = "----TurboDocxSdk#{SecureRandom.hex(16)}"

      body = build_multipart_body(form_data, boundary)

      klass = method == "PATCH" ? Net::HTTP::Patch : Net::HTTP::Post
      request = klass.new(url)
      request["Content-Type"] = "multipart/form-data; boundary=#{boundary}"
      if @access_token
        request["Authorization"] = "Bearer #{@access_token}"
      elsif @api_key
        request["Authorization"] = "Bearer #{@api_key}"
      end
      request["x-rapiddocx-org-id"] = @org_id if @org_id
      request.body = body

      execute_multipart(request, url)
    rescue TurboDocxError
      raise
    rescue StandardError => e
      raise NetworkError, "Form data request failed: #{e.message}"
    end

    def build_multipart_body(form_data, boundary)
      parts = []
      form_data.each do |key, value|
        if value.is_a?(Array)
          # Multiple parts under the same field name (e.g. multiple images)
          value.each { |item| append_part(parts, key, item, boundary) }
        else
          append_part(parts, key, value, boundary)
        end
      end
      parts << "--#{boundary}--\r\n"
      parts.join
    end

    def append_part(parts, key, value, boundary)
      if value.is_a?(Hash) && value.key?(:io)
        # File part
        parts << "--#{boundary}\r\n"
        parts << "Content-Disposition: form-data; name=\"#{key}\"; filename=\"#{value[:filename]}\"\r\n"
        parts << "Content-Type: #{value[:content_type]}\r\n\r\n"
        io = value[:io]
        io.rewind if io.respond_to?(:rewind)
        parts << io.read
        parts << "\r\n"
      else
        # Text part
        parts << "--#{boundary}\r\n"
        parts << "Content-Disposition: form-data; name=\"#{key}\"\r\n\r\n"
        parts << value.to_s
        parts << "\r\n"
      end
    end

    def execute_multipart(request, url)
      response = http_request(request, url)
      handle_error_response(response) unless response.is_a?(Net::HTTPSuccess)
      json_data = JSON.parse(response.body)
      ResponseNormalizer.normalize(smart_unwrap(json_data))
    end

    # Smart unwrap: if response has ONLY a "data" key, extract its value.
    def smart_unwrap(data)
      if data.is_a?(Hash)
        keys = data.keys
        if keys.length == 1 && keys[0] == "data"
          return data["data"]
        end
      end
      data
    end

    def handle_error_response(response)
      message = "HTTP #{response.code}: #{response.message}"
      begin
        error_data = JSON.parse(response.body)
        message = error_data["message"] || error_data["error"] || message
      rescue StandardError
        # If response is not JSON, use status text
      end

      code = response.code.to_i
      case code
      when 400 then raise ValidationError, message
      when 401 then raise AuthenticationError, message
      when 404 then raise NotFoundError, message
      when 429 then raise RateLimitError, message
      else raise TurboDocxError.new(message, status_code: code)
      end
    end

    def detect_file_type(content)
      bytes = content.bytes
      # PDF: %PDF
      if bytes[0] == 0x25 && bytes[1] == 0x50 && bytes[2] == 0x44 && bytes[3] == 0x46
        return "application/pdf"
      end
      # ZIP-based (DOCX, PPTX): PK
      if bytes[0] == 0x50 && bytes[1] == 0x4B
        snippet = content[0, [content.length, 2000].min]
        if snippet.include?("ppt/")
          return "application/vnd.openxmlformats-officedocument.presentationml.presentation"
        end
        return "application/vnd.openxmlformats-officedocument.wordprocessingml.document"
      end
      "application/octet-stream"
    end
  end
end
